### PR TITLE
Feat/sign2

### PR DESCRIPTION
**Pull request by dupuchba**
Signing a request with the v2 protocol
Working BUT
Version is hardcoded and `query` protocol generation has not been done yet so it's not tested on all services

### CODE REVIEW

File: src/portkey/awssig.clj [hunk 1]
```diff
@@ -126,3 +127,75 @@
...
+        (loop [i 0]
+          (if (< i n)
+            (let [r (- (aget a i) (aget b i))]
+               (if (zero? r)
+                 (recur (inc i))
+                 r))
+            (- (alength a) (alength b))))))
+
+(defn- create-canonicalized-query-string-v2
+   "Step 1 : Create a canonicalized query-string for building the signature v2."
```
Comment (viesti) [inline src/portkey/awssig.clj]: I'd leave out the order of use of signature helper functions out of docstrings, order of steps should be visible from place of use already, right?

File: src/portkey/awssig.clj [hunk 2]
```diff
@@ -126,3 +127,75 @@
...
+   [& {:keys [query-params]}]
+   (x/str
+     (comp
+       (x/sort-by key utf8-natural-byte-cmp)
+       (x/for [[k v] %]
+              (str (codec/url-encode k) "=" (codec/url-encode v)))
```
Comment (viesti) [inline src/portkey/awssig.clj]: Could [URLEncoder](https://docs.oracle.com/javase/8/docs/api/java/net/URLEncoder.html) be used here (would drop the dependency to ring-codec)?

File: src/portkey/awssig.clj [hunk 3]
```diff
@@ -126,3 +127,75 @@
...
+       (interpose "&"))
+     query-params))
+
+(defn- create-string-to-sign-v2
+   "Step 2 : Gather request information before signing."
+   [verb host uri qs]
+   (->> [verb "\n" host "\n" uri "\n" qs]
```
Comment (viesti) [inline src/portkey/awssig.clj]: Maybe just `(str verb "\n" host "\n" uri "\n" qs)`

File: src/portkey/awssig.clj (hunk 1 above)
Comment (dupuchba) [reply src/portkey/awssig.clj]: Yes indeed but I thought repeating doc's steps was easier to understand :-)

File: src/portkey/awssig.clj (hunk 2 above)
Comment (dupuchba) [reply src/portkey/awssig.clj]: Yes, I'll do that.

File: src/portkey/awssig.clj (hunk 3 above)
Comment (dupuchba) [reply src/portkey/awssig.clj]: Yes that's a good one 👍

File: src/portkey/awssig.clj (hunk 2 above)
Comment (cgrand) [reply src/portkey/awssig.clj]: Don't forget to the check the hexadecimal digits produced by the encoder are all caps (as required for canonicalization)

File: src/portkey/awssig.clj [hunk 4]
```diff
@@ -126,3 +127,75 @@
...
+(def x-amz-date-formatter-sign-2
+  (java.time.format.DateTimeFormatter/ofPattern "yyyy-MM-dd'T'HH:mm:ss'Z'"))
+
+(defn- hmac-sha-256-sign2
+   [^bytes secret ^String s]
+   (-> (javax.crypto.Mac/getInstance "HmacSHA256")
+         (doto (.init (javax.crypto.spec.SecretKeySpec. secret "HmacSHA256")))
+         (.doFinal (.getBytes s "UTF8"))))
+
+(defn- utf8-natural-byte-cmp
```
Comment (cgrand) [inline src/portkey/awssig.clj]: This requirement felt bizarre initially but now I understand it: it's the simple unambiguous order one can implement over strings. Default orders are often dependent on the (programming) language, locale, env (OS etc.).

File: src/portkey/awssig.clj (hunk 1 above)
Comment (cgrand) [reply src/portkey/awssig.clj]: While I agree with @viesti, these are private vars so it doesn't bother me.

Comment (dupuchba) [reply src/portkey/awssig.clj]: as you prefer for this one

File: src/portkey/awssig.clj (hunk 2 above)
Comment (dupuchba) [reply src/portkey/awssig.clj]: Ok

File: src/portkey/awssig.clj (hunk 4 above)
Comment (dupuchba) [reply src/portkey/awssig.clj]: OK, I keep it like that.

**Comment by dupuchba** [comment]
I'll do all the change this weekend as I am working on this project saturday & sunday, really excited to see if I can generate some query requests ^^

File: src/portkey/awssig.clj (hunk 1 above)
Comment (cgrand) [reply src/portkey/awssig.clj]: Principle of least effort: don’t change anything! Kimmo, any strong argument is against?

Comment (viesti) [reply src/portkey/awssig.clj]: True. I'm ok with leaving the steps in docstrings, not necessary to do more effort now.

**Comment by dupuchba** [comment]
You will find my new commit concerning generating namespaces for the `query` protocol.

#### Where am I :
- OK for Amazon SimpleDB with signing v2
- I have to try for the other domains

#### What is left :
- Figuring out how to use the `move` parameters (used for 3/4 functions in the `query` protocol
- Output coercion : it’s XML right now
- Passing the version date information to other namespace (not AWS SimpleDB)
- Generating the right payload for Attributes.Name.x = 

#### Points where I am not confortable yet
- `gen-operation` & `gen-operation-query` use a lot of the same code, that’s not viable 
- I had to reload manually all middleware in `sync-http-client` in order to fully signv2 the request because `wrap-sign` was call last, with body already coerced
- globally I think we can reduce my code greatly

#### Next steps : 
- correct comments on this PR
- Do the what’s left and repeat :D

/cc @cgrand @viesti

File: src/portkey/awssig.clj (hunk 2 above)
Comment (viesti) [reply src/portkey/awssig.clj]: In following principle of least effort, I'd be ok with leaving `ring-codec` dependency, we can drop it later.

File: src/portkey/aws.clj [hunk 5]
```diff
@@ -103,7 +103,33 @@
       (apply client req args))))
 
 (defn sync-http-client [req coerce-resp]
-  (http/with-additional-middleware [wrap-sign]
```
Comment (cgrand) [inline src/portkey/aws.clj]: I came to hate middlewares.

**Comment by dupuchba** [comment]
Added test for signing v4 from amazon as there are some corner cases not working when signing v4 with the query protocol. It is still a wip.
I'll probably have to split the `sign4` functions to make the test easier/pass
